### PR TITLE
Add syevd LAPACK Hermitian eigensolver 

### PR DESCRIPTION
**Pull request by jaemolihm**
Added `syevd!` wrapper to LAPACK complex Hermitian eigensolvers zheevd and cheevd.

It is more accurate than the existing wrappers (`syev!`, `syevr`), and is also the fastest for large matrices (N > 100) with MKL.

https://netlib.org/lapack/lawnspdf/lawn183.pdf
> QR [SYEV] and DC [SYEVD] are the most accurate algorithms, measured both in terms of producing pairwise orthogonal eigenvectors and small residuals norms. MR [SYEVR] is less accurate but still achieves errors of size O(nε), where n is the dimension and ε is machine epsilon

This is useful e.g. in https://github.com/JuliaMolSim/DFTK.jl/pull/842

**Benchmark (MKL, N = 100)**
| Method | Timing (ms) | Allocation (KiB) | Orthogonality error |
|:------:|:-----------:|:----------:|---------------------|
|  syev! | 1.808       | 36.36      | 2.3e-14             |
| syevr! | 1.577       | 394.62     | 1.5e-13             |
| syevd! | 1.641       | 373.23     | 1.9e-14             |

**Benchmark (MKL, N = 1000)**
| Method | Timing (ms) | Allocation (MiB) | Orthogonality error |
|:------:|:-----------:|:----------:|---------------------|
|  syev! | 793       | 0.5      | 2.4e-13             |
| syevr! | 547      | 31.3     | 5.2e-12             |
| syevd! | 504       | 31.1     | 9.9e-14             |

<details>
<summary>Benchmark code</summary>
```julia
using LinearAlgebra
using BenchmarkTools
using MKL

function test_syev(h, tmp, jobz='V')
  tmp .= h
  LinearAlgebra.LAPACK.syev!(jobz, 'U', tmp)
end
function test_syevr(h, tmp, jobz='V')
  tmp .= h
  LinearAlgebra.LAPACK.syevr!(jobz, tmp)
end
function test_syevd(h, tmp, jobz='V')
  tmp .= h
  LinearAlgebra.LAPACK.syevd!(jobz, 'U', tmp)
end

N = 1000
T = Float64
h = rand(Complex{T}, N, N);
h .+= h';
tmp = zeros(Complex{T}, N, N);

@btime test_syev($h, $tmp);
@btime test_syevr($h, $tmp);
@btime test_syevd($h, $tmp);

e, v = LinearAlgebra.LAPACK.syev!('V', 'U', copy(h));
@info norm(v' * v - I)
e, v = LinearAlgebra.LAPACK.syevr!('V', copy(h));
@info norm(v' * v - I)
e, v = LinearAlgebra.LAPACK.syevd!('V', 'U', copy(h));
@info norm(v' * v - I)
```
</details>

### CODE REVIEW

**Review by ViralBShah — APPROVED** [review]

**Comment by dkarrasch** [comment]
Cool. Are there cases where we want to dispatch to this method by default, or do we just want to provide this?

**Comment by ViralBShah** [comment]
@jaemolihm The tests introduced for this PR are failing.

**Comment by jaemolihm** [comment]
Thank you, I fixed the test by adding the real symmetric eigensolver interface.

> Cool. Are there cases where we want to dispatch to this method by default, or do we just want to provide this?

The motivation is just to provide this. Choosing the default is more difficult because the timing also depends on the library used (and possibly on the hardware).

**Comment by ViralBShah** [comment]
Is it worth considering providing an option to the eigensolver interface to use this? Could be in a separate PR.

**Comment by jaemolihm** [comment]
> Is it worth considering providing an option to the eigensolver interface to use this? Could be in a separate PR.

Yes, it seems useful. Maybe something like below?

```julia
using LinearAlgebra.LAPACK: syev!, syevd!, syevr!
using LinearAlgebra: RealHermSymComplexHerm, eigencopy_oftype, BlasReal, Eigen, sorteig!, LAPACK, eigtype
import LinearAlgebra: eigen!, eigen

function eigen(A::RealHermSymComplexHerm, method::Symbol; sortby::Union{Function,Nothing}=nothing)
    S = eigtype(eltype(A))
    eigen!(eigencopy_oftype(A, S), method; sortby)
end

function eigen!(A::RealHermSymComplexHerm{<:BlasReal,<:StridedMatrix}, method::Symbol; sortby::Union{Function,Nothing}=nothing)
    if method == :syevr
      Eigen(sorteig!(LAPACK.syevr!('V', 'A', A.uplo, A.data, 0.0, 0.0, 0, 0, -1.0)..., sortby)...)
    elseif method == :syev
      Eigen(sorteig!(LAPACK.syev!('V', A.uplo, A.data)..., sortby)...)
    elseif method == :syevd
      Eigen(sorteig!(LAPACK.syevd!('V', A.uplo, A.data)..., sortby)...)
    else
      throw(ArgumentError("Wrong method $method. Must be :syevr or :syev or :syevd."))
    end
end
```

**Comment by ViralBShah** [comment]
Yes, something along those lines, but perhaps with slightly more descriptive names?

File: stdlib/LinearAlgebra/src/lapack.jl [hunk 1]
```diff
@@ -5504,6 +5589,20 @@ The eigenvalues are returned in `W` and the eigenvectors in `Z`.
...
+    syevd!(jobz, uplo, A)
+
+Finds the eigenvalues (`jobz = N`) or eigenvalues and eigenvectors
+(`jobz = V`) of a symmetric matrix `A`. If `uplo = U`, the upper triangle
+of `A` is used. If `uplo = L`, the lower triangle of `A` is used.
+
+Use the divide-and-conquer method, instead of the QR iteration used by
+`syev!` or multiple relatively robust representations used by `syevr!`.
+See James W. Demmel et al, SIAM J. Sci. Comput. 30, 3, 1508 (2008) for
+a comparison of the accuracy and performatce of different methods.
```
[the remark this answers is not in the document]
Comment (jaemolihm) [reply stdlib/LinearAlgebra/src/lapack.jl:5602]: I fixed this typo in the follow up PR #49355.

**Pull request merged**